### PR TITLE
Improve Error Handling

### PR DESCRIPTION
**Pull request by hug33k**
This PR adds a new package : `errors`

This package has a function called `GetErrorMessage` that will generate a user-friendly message for an error.

Two use cases :
- You got an error from an external package : You can add the error type into the switch case and choose how to handle the error.
- You need to create a custom error : You can create a new error type into this package and add it to the switch case.

### CODE REVIEW

File: costs/anomalies/anomalies_route.go [hunk 1]
```diff
@@ -81,11 +82,21 @@ func makeElasticSearchRequest(ctx context.Context, parsedParams AnomalyEsQueryPa
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/anomalies/anomalies_route.go]: Use the `data` param to pass the error. 
`l.Error("...", map[string]interface{}{"error": err.Error()})`

File: costs/anomalies/anomalies_route.go [hunk 2]
```diff
@@ -81,11 +82,21 @@ func makeElasticSearchRequest(ctx context.Context, parsedParams AnomalyEsQueryPa
...
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline costs/anomalies/anomalies_route.go]: Are you sure you want tell the user such an accurate error? ElasticSearch should not appear to me.

File: costs/costs_route.go [hunk 3]
```diff
@@ -120,11 +121,21 @@ func makeElasticSearchRequestAndParseIt(ctx context.Context, parsedParams esQuer
...
+			return es.SimplifiedCostsDocument{}, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return es.SimplifiedCostsDocument{}, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
+			return es.SimplifiedCostsDocument{}, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline costs/costs_route.go]: Same as above.

File: costs/costs_route.go [hunk 4]
```diff
@@ -120,11 +121,21 @@ func makeElasticSearchRequestAndParseIt(ctx context.Context, parsedParams esQuer
...
+			})
+			return es.SimplifiedCostsDocument{}, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return es.SimplifiedCostsDocument{}, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/costs_route.go]: Same as above.

File: costs/diff/diff.go [hunk 5]
```diff
@@ -18,11 +18,11 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/trackit/trackit-server/errors"
```
Comment (thibautcornolti) [inline costs/diff/diff.go]: This import should not be here

File: costs/diff/diff_route.go [hunk 6]
```diff
@@ -97,11 +98,21 @@ func makeElasticSearchRequest(ctx context.Context, parsedParams esQueryParams) (
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/diff/diff_route.go]: Same as above.

File: costs/diff/diff_route.go [hunk 7]
```diff
@@ -97,11 +98,21 @@ func makeElasticSearchRequest(ctx context.Context, parsedParams esQueryParams) (
...
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline costs/diff/diff_route.go]: Same as above.

File: costs/tags/tags_keys.go [hunk 8]
```diff
@@ -51,12 +52,12 @@ func getTagsKeysWithParsedParams(ctx context.Context, params tagsKeysQueryParams
 		if returnCode == http.StatusOK {
 			return returnCode, response
 		}
-		return returnCode, errors.New("Internal server error")
+		return returnCode, err
```
Comment (thibautcornolti) [inline costs/tags/tags_keys.go]: Why didn't you use GetErrorMessage here?

File: costs/tags/tags_keys.go [hunk 9]
```diff
@@ -81,11 +82,21 @@ func makeElasticSearchRequestForTagsKeys(ctx context.Context, params tagsKeysQue
...
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline costs/tags/tags_keys.go]: Same as above.

File: costs/tags/tags_keys.go [hunk 10]
```diff
@@ -81,11 +82,21 @@ func makeElasticSearchRequestForTagsKeys(ctx context.Context, params tagsKeysQue
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/tags/tags_keys.go]: Same as above.

File: costs/tags/tags_values.go [hunk 11]
```diff
@@ -88,12 +89,12 @@ func getTagsValuesWithParsedParams(ctx context.Context, params tagsValuesQueryPa
 		if returnCode == http.StatusOK {
 			return returnCode, response
 		}
-		return returnCode, errors.New("Internal server error")
+		return returnCode, err
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Same as above.

File: costs/tags/tags_values.go [hunk 12]
```diff
@@ -143,11 +144,21 @@ func makeElasticSearchRequestForTagsValues(ctx context.Context, params tagsValue
...
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Same as above.

File: costs/tags/tags_values.go [hunk 13]
```diff
@@ -143,11 +144,21 @@ func makeElasticSearchRequestForTagsValues(ctx context.Context, params tagsValue
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline costs/tags/tags_values.go]: Same as above.

File: errors/database.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/trackit/jsonlog"
+)
+
+const DatabaseGenericError = -1
+const DatabaseItemNotFound = 0
```
Comment (thibautcornolti) [inline errors/database.go]: Excellent context to use `Iota` https://golang.org/ref/spec#Iota

File: errors/shared_account.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/trackit/jsonlog"
+)
+
+const SharedAccountGenericError = -1
+const SharedAccountNoPermission = 0
+const SharedAccountBadPermission = 1
+const SharedAccountRequestError = 2
```
Comment (thibautcornolti) [inline errors/shared_account.go]: Same here, perfect context to use `Iota`

File: s3/costs/s3_costs_route.go [hunk 16]
```diff
@@ -18,17 +18,19 @@ import (
 	"context"
 	"database/sql"
 	"fmt"
+	"github.com/trackit/trackit-server/errors"
```
Comment (thibautcornolti) [inline s3/costs/s3_costs_route.go]: This import should not be here.

File: usageReports/ec2/get_ec2_data.go [hunk 17]
```diff
@@ -100,11 +101,21 @@ func makeElasticSearchEc2MonthlyRequest(ctx context.Context, parsedParams ec2Que
...
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline usageReports/ec2/get_ec2_data.go]: Same as above.

File: usageReports/ec2/get_ec2_data.go [hunk 18]
```diff
@@ -100,11 +101,21 @@ func makeElasticSearchEc2MonthlyRequest(ctx context.Context, parsedParams ec2Que
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline usageReports/ec2/get_ec2_data.go]: Same as above.

File: usageReports/ec2/prepare_response.go [hunk 19]
```diff
@@ -17,6 +17,7 @@ package ec2
 import (
 	"context"
 	"encoding/json"
+	"github.com/trackit/trackit-server/errors"
```
Comment (thibautcornolti) [inline usageReports/ec2/prepare_response.go]: This import should not be here.

File: usageReports/rds/get_rds_data.go [hunk 20]
```diff
@@ -46,11 +47,21 @@ func makeElasticSearchCostRequest(ctx context.Context, params rdsQueryParams) (*
...
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline usageReports/rds/get_rds_data.go]: Same as above.

File: usageReports/rds/get_rds_data.go [hunk 21]
```diff
@@ -46,11 +47,21 @@ func makeElasticSearchCostRequest(ctx context.Context, params rdsQueryParams) (*
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline usageReports/rds/get_rds_data.go]: Same as above.

File: usageReports/rds/get_rds_data.go [hunk 22]
```diff
@@ -100,11 +111,21 @@ func makeElasticSearchRdsMonthlyRequest(ctx context.Context, parsedParams rdsQue
...
+			return nil, http.StatusInternalServerError, fmt.Errorf("could not execute the ElasticSearch query")
```
Comment (thibautcornolti) [inline usageReports/rds/get_rds_data.go]: Same as above.

File: usageReports/rds/get_rds_data.go [hunk 23]
```diff
@@ -100,11 +111,21 @@ func makeElasticSearchRdsMonthlyRequest(ctx context.Context, parsedParams rdsQue
...
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
+			})
+			return nil, http.StatusInternalServerError, errors.GetErrorMessage(ctx, err)
+		} else {
+			l.Error("Query execution failed : "+err.Error(), nil)
```
Comment (thibautcornolti) [inline usageReports/rds/get_rds_data.go]: Same as above.

File: usageReports/rds/prepare_response.go [hunk 24]
```diff
@@ -17,6 +17,7 @@ package rds
 import (
 	"context"
 	"encoding/json"
+	"github.com/trackit/trackit-server/errors"
```
Comment (thibautcornolti) [inline usageReports/rds/prepare_response.go]: This import should not be here.

**Review by thibautcornolti — COMMENTED** [review]
I still think the error messages are slightly too accurate for a customer, especially on ES errors.

**Review by thibautcornolti — CHANGES_REQUESTED** [review]
Changes requested in the previous comments

File: costs/tags/tags_keys.go (hunk 8 above)
Comment (hug33k) [reply costs/tags/tags_keys.go]: `GetErrorMessage`is already called in `makeElasticSearchRequestForTagsKeys`

File: costs/tags/tags_values.go (hunk 11 above)
Comment (hug33k) [reply costs/tags/tags_values.go]: `GetErrorMessage` is already called in `makeElasticSearchRequestForTagsKeys`

File: costs/anomalies/anomalies_route.go [hunk 25]
```diff
@@ -81,11 +82,20 @@ func makeElasticSearchRequest(ctx context.Context, parsedParams AnomalyEsQueryPa
...
-			return nil, http.StatusOK, err
+			l.Warning("Query execution failed, ES index does not exists", map[string]interface{}{
+				"index": index,
+				"error": err.Error(),
+			})
+			return nil, http.StatusOK, errors.GetErrorMessage(ctx, err)
+		} else if err.(*elastic.Error).Details.Type == "search_phase_execution_exception" {
+			l.Error("Error while getting data from ES", map[string]interface{}{
+				"type": fmt.Sprintf("%T", err),
+				"error": err,
```
Comment (simonmeyerrr) [inline costs/anomalies/anomalies_route.go:93]: shouldn't it be `err.Error()` ? (this case appears many times all over your code)

File: errors/database.go [hunk 26]
```diff
@@ -0,0 +1,47 @@
+package errors
```
Comment (simonmeyerrr) [inline errors/database.go:15]: You forgot the TrackIt header.

File: errors/errors.go [hunk 27]
```diff
@@ -0,0 +1,80 @@
+package errors
```
Comment (simonmeyerrr) [inline errors/errors.go:15]: same

File: errors/shared_account.go [hunk 28]
```diff
@@ -0,0 +1,61 @@
+package errors
```
Comment (simonmeyerrr) [inline errors/shared_account.go:15]: same

**Review by simonmeyerrr — CHANGES_REQUESTED** [review]

File: costs/anomalies/anomalies_route.go (hunk 25 above)
Comment (hug33k) [reply costs/anomalies/anomalies_route.go:93]: `err` may have more information than just an error message, so we need to see the entire structure.

**Review by simonmeyerrr — APPROVED** [review]

**Review by thibautcornolti — APPROVED** [review]

**Pull request merged**